python/src/aica_api/client.py

```diff
@@ -54,7 +59,56 @@ def _endpoint(self, endpoint=''):
         """
         if self._protocol is None:
             self.protocol()
-        return f'{self._address}/{self._protocol}/{endpoint}'
+        return self.__raw_endpoint(f'{self._protocol}/{endpoint}')
+
+    def __raw_endpoint(self, endpoint: str) -> str:
+        return f'{self._address}/{endpoint}'
+
+    def __create_token(self) -> None:
+        """Authenticate with the API and store the result in self.__token."""
+        if self.__token is None:
+            return
+        rep = requests.post(self._endpoint('auth/login'), headers={'Authorization': f'Bearer {self.__api_key}'})
+        rep.raise_for_status()
+        self.__token = rep.json()['token']
+
+    def _sio_auth(self) -> Optional[str]:
+        # FIXME: doesn't handle token expiration
+        if self.__api_key is not None:
+            self.__create_token()
+        return self.__token
+
+    def _safe_uri(self, uri: str) -> str:
+        """
+        Make a string safe for use in a URI by encoding special characters.
+
+        :param uri: The URI to sanitize
+        :return: The sanitized URI
+        """
+        return urllib.parse.quote_plus(uri)
+
+    def _request(
+        self,
+        method: str,
+        endpoint: str,
+        *,
+        params: Optional[dict] = None,
+        auth: Optional[bool] = None,
+        json: Optional[dict] = None,
+    ) -> requests.Response:
+        headers = None
+        retry = 2
+        while retry > 0:
+            if self.__api_key is not None:
+                self.__create_token()
+                headers = {'Authorization': f'Bearer {self.__token}'}
+            res = requests.request(method, self._endpoint(endpoint), params=params, json=json, headers=headers, timeout=5)
+            retry -= 1
+            if res.status_code == 401:
+                if self.__api_key is None:
```

Comment:
is there a `not` missing or could you explain me how this function works?

Reply:
If you have an API key:

* Ensure your token exists and use it in auth
* Do your request
* If it's 401, delete the token
* Retry, recreating the token (as it is now `None`)

This ensures that we retry on a 401 failure, which should mean that our token has expired and need to be recreated (i.e. first step the token existed and we just reused it).

If you don't an API key:

* Do your request
* If it's 401, exit

No need to retry anything, it won't change the outcome.